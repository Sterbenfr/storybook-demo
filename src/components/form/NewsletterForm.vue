<template>
  <form @submit.prevent="handleSubmit">
    <InputField v-model="formData.name" label="Name" :error="errors.name" />
    <InputField v-model="formData.email" label="Email" type="email" :error="errors.email" />
    <SubmitButton :loading="loading">Submit</SubmitButton>
  </form>
</template>

<script setup lang="ts">
import { ref } from 'vue'
import InputField from '../form-components/InputField.vue'
import SubmitButton from '../form-components/SubmitButton.vue'

/**
 * Form data object
 */
const formData = ref({
  name: '',
  email: '',
})

/**
 * Error messages for form fields
 */
const errors = ref({
  name: '',
  email: '',
})

/**
 * Loading state for form submission
 */
const loading = ref(false)

/**
 * Validate the form fields
 * @returns {boolean} - Returns true if the form is valid, otherwise false
 */
const validateForm = () => {
  let valid = true
  if (!formData.value.name) {
    errors.value.name = 'Name is required'
    valid = false
  } else {
    errors.value.name = ''
  }

  if (!formData.value.email) {
    errors.value.email = 'Email is required'
    valid = false
  } else if (!/\S+@\S+\.\S+/.test(formData.value.email)) {
    errors.value.email = 'Email is invalid'
    valid = false
  } else {
    errors.value.email = ''
  }

  return valid
}

/**
 * Handle form submission
 */
const handleSubmit = () => {
  if (validateForm()) {
    loading.value = true
    // Simulate form submission
    setTimeout(() => {
      loading.value = false
      alert('Form submitted!')
    }, 1000)
  }
}
</script>
