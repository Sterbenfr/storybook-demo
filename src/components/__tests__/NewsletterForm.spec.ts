import { describe, it, expect, beforeEach, vi } from 'vitest'
import { mount } from '@vue/test-utils'
import NewsletterForm from '../form/NewsletterForm.vue'

// Mock the v-text-field component
const VTextField = {
  template: `
    <div>
      <input :value="modelValue" @input="$emit('update:modelValue', $event.target.value)" />
      <div v-if="errorMessages.length" class="v-messages__message">{{ errorMessages[0] }}</div>
    </div>
  `,
  props: ['modelValue', 'label', 'type', 'errorMessages', 'outlined', 'dense', 'hideDetails'],
}

// Mock window.alert
global.alert = vi.fn()

describe('NewsletterForm.vue', () => {
  let wrapper: any

  beforeEach(() => {
    wrapper = mount(NewsletterForm, {
      global: {
        components: {
          'v-text-field': VTextField,
        },
      },
    })
  })

  it('renders correctly', () => {
    expect(wrapper.exists()).toBe(true)
  })

  it('matches snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('updates form data when input fields change', async () => {
    const inputs = wrapper.findAll('input')
    const nameInput = inputs[0]
    const emailInput = inputs[1]

    await nameInput.setValue('John Doe')
    await emailInput.setValue('john.doe@example.com')

    expect(wrapper.vm.formData.name).toBe('John Doe')
    expect(wrapper.vm.formData.email).toBe('john.doe@example.com')
  })

  it('displays validation errors when fields are invalid', async () => {
    await wrapper.find('form').trigger('submit.prevent')

    expect(wrapper.find('.v-messages__message').exists()).toBe(true)
    expect(wrapper.find('.v-messages__message').text()).toBe('Name is required')
  })

  it('submits the form when fields are valid', async () => {
    wrapper.vm.formData = {
      name: 'John Doe',
      email: 'john.doe@example.com',
    }

    await wrapper.find('form').trigger('submit.prevent')

    expect(wrapper.emitted()).toHaveProperty('submit')
  })

  it('handles loading state correctly', async () => {
    wrapper.vm.formData = {
      name: 'John Doe',
      email: 'john.doe@example.com',
    }

    await wrapper.find('form').trigger('submit.prevent')

    expect(wrapper.vm.loading).toBe(true)

    // Simulate form submission delay
    await new Promise((resolve) => setTimeout(resolve, 1000))

    expect(wrapper.vm.loading).toBe(false)
  })
})
