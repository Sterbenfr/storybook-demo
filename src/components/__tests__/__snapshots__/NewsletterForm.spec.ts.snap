// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`NewsletterForm.vue > matches snapshot 1`] = `
"<form>
  <div data-v-901038f8="" class="input-wrapper">
    <div data-v-901038f8="" class="custom-text-field"><input value="">
      <!--v-if-->
    </div>
  </div>
  <div data-v-901038f8="" class="input-wrapper">
    <div data-v-901038f8="" class="custom-text-field"><input value="">
      <!--v-if-->
    </div>
  </div><button data-v-af806622="" type="submit"><span data-v-af806622="">Submit</span></button>
</form>"
`;
